src/java_bytecode/java_types.cpp

```diff
@@ -167,6 +167,218 @@ exprt java_bytecode_promotion(const exprt &expr)
+                     INVARIANT(is_reference(type),
+                               "All generic parameters should be references");
+                     java_generic_inst_parametert inst_type(
+                         to_symbol_type(type.subtype()));
+                     return inst_type;
+                   }
+                 });
+}
+
+/// Take a signature string and remove everything in angle brackets allowing
+/// for the type to be parsed normally.
+/// \param src: The input string
+/// \return The input string with everything between angle brackets removed
+std::string erase_type_arguments(const std::string &src) {
+  std::string class_name = src;
+  std::size_t f_pos = class_name.find('<', 1);
+
+  while (f_pos != std::string::npos) {
+    std::size_t e_pos = find_closing_delimiter(class_name, f_pos, '<', '>');
+    if (e_pos == std::string::npos)
+      throw unsupported_java_class_signature_exceptiont("recursive generic");
+
+    // erase generic information between brackets
+    class_name.erase(f_pos, e_pos - f_pos + 1);
+
+    // Search the remainder of the string for generic signature
+    f_pos = class_name.find('<', e_pos + 1);
+  }
+  return class_name;
+}
+
+/// Returns the full class name, skipping over the generics
+/// \param src: a type descriptor or signature like LOuterClass<TT;>.Inner;
+/// \return The full name of the class like OuterClass.Inner
+std::string gather_full_class_name(const std::string &src) {
+  PRECONDITION(src[0] == 'L');
+  PRECONDITION(src[src.size() - 1] == ';');
+
+  std::string class_name = src.substr(1, src.size() - 2);
+
+  class_name = erase_type_arguments(class_name);
+
+  std::replace(class_name.begin(), class_name.end(), '.', '$');
+  std::replace(class_name.begin(), class_name.end(), '/', '.');
+  return class_name;
+}
+
+size_t find_closing_semi_colon_for_reference_type(const std::string src,
+                                                  size_t starting_point = 0);
+
+/// Given a substring of a descriptor or signature that contains 1 or more types
+/// parse out the inidividual types. This is used for parsing the parameters of
+/// a function or the gneeric arguments contained within angle brackets.
+/// \param src: The input string that is wrapped in either ( ) or < >
+/// \param class_name_prefix: The name of the class to use to prefix any found
+///   generic types
+/// \param opening_bracket: For checking string is passed in as expected, the
+///   opening bracket (i.e. '(' or '<').
+/// \param closing_bracket: For checking string is passed in as expected, the
+///   closing bracket (i.e. ')' or '>').
+/// \return A vector of types that the string represents
+std::vector<typet> parse_list_types(const std::string src,
+                                    const std::string class_name_prefix,
+                                    const char opening_bracket,
+                                    const char closing_bracket) {
+  PRECONDITION(src[0] == opening_bracket);
```

Comment:
Before these: precondition src.size()>=2

Reply:
Also done for other functions with similar preconditions